python/taichi/lang/common_ops.py

```diff
@@ -1,61 +1,135 @@
+def numpy_or_constant(x):
+    import taichi as ti
+    if ti.is_taichi_class(x):
+        return x.to_numpy()
+    else:
+        return x
+
+
 class TaichiOperations:
     def __neg__(self):
+        if self.is_pyconstant():
+            return self.make_from_numpy(-self.to_numpy())
+
         import taichi as ti
         return ti.neg(self)
 
     def __abs__(self):
+        if self.is_pyconstant():
+            return self.make_from_numpy(np.abs(self.to_numpy()))
+
         import taichi as ti
         return ti.abs(self)
 
     def __add__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(self.to_numpy() +
+                                        numpy_or_constant(other))
+
         import taichi as ti
         return ti.add(self, other)
 
     def __radd__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(
+                numpy_or_constant(other) + self.to_numpy())
+
         import taichi as ti
         return ti.add(other, self)
 
     def __sub__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(self.to_numpy() -
+                                        numpy_or_constant(other))
+
         import taichi as ti
         return ti.sub(self, other)
 
     def __rsub__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(
+                numpy_or_constant(other) - self.to_numpy())
+
         import taichi as ti
         return ti.sub(other, self)
 
     def __mul__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(self.to_numpy() *
+                                        numpy_or_constant(other))
+
         import taichi as ti
         return ti.mul(self, other)
 
     def __rmul__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(
+                numpy_or_constant(other) * self.to_numpy())
+
         import taichi as ti
         return ti.mul(other, self)
 
     def __truediv__(self, other):
+        if self.is_pyconstant():
+            return self.make_from_numpy(self.to_numpy() /
+                                        numpy_or_constant(other))
+
```

Comment:
I feel like invoking numpy array operations here is slightly cumbersome, also the code here is repeating itself. Maybe a better solution is to further decorate operators such as `ti.truediv` to let them treat the case where elements of a matrix are purely constants.

Reply:
> Maybe a better solution is to further decorate operators such as ti.truediv to let them treat the case where elements of a matrix are purely constants.

That sounds a gret solution to make the code less repeating itself. Let's resolve it in a spearate PR? This PR is fat and stale.